### ISSUE TITLE
BUG/API: index.array for RangeIndex, MultiIndex

### ISSUE THREAD

**Issue by jbrockmendel** [pandas-dev/pandas#37277]
ATM MultiIndex.array raises ValueError, since there is no single array actually backing MI.  RangeIndex creates and returns a PandasArray.  I think RangeIndex should also raise (at least until/unless we implement a RangeArray).

This also has implications for extract_array, which currently raises on MultiIndex, which has led to some ugly checks in e.g. Index._cmp_method.

**Comment by jbrockmendel** [comment]
@TomAugspurger @jorisvandenbossche thoughts here on adding something like a `preserve_mi=False` kwarg to extract_array to return unchanged instead of raising?

**Comment by jorisvandenbossche** [comment]
I don't have a strong opinion about it, but so that would basically the same as:

```
arr = ...

if not isinstance(arr, MultiIndex):
    arr = extract_array(arr)
```

So it would save one line of boiler plate in every place where it is needed

**Comment by TomAugspurger** [comment]
Agreed that extract_array shouldn't materialize a PandasArray from a
RangeIndex. IMO, it should only... extract, existing arrays. Not create new
ones :)
